router/core/modules_v1.go

```diff
@@ -0,0 +1,166 @@
+	m := instance.MyModule()
+
+	if m.ID == "" {
+		panic("MyModule.ID is required")
+	}
+	if val := m.New(); val == nil {
+		panic("MyModuleInfo.New must return a non-nil module instance")
+	}
+
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	if _, ok := r.modules[m.ID]; ok {
+		panic(fmt.Sprintf("MyModule already registered: %s", m.ID))
+	}
+	r.modules[m.ID] = m
+}
+
+// sortMyModules sorts the modules by priority, 0 is the highest priority, is the first to be executed.
+// If two modules have the same priority, they are sorted by registration order.
+// If a module has no priority, it is considered to have the lowest priority.
+func sortMyModules(modules []MyModuleInfo) []MyModuleInfo {
+	sort.Slice(modules, func(i, j int) bool {
+		var priorityI, priorityJ int = math.MaxInt, math.MaxInt
+		if modules[i].Priority != nil {
+			priorityI = *modules[i].Priority
+		}
+		if modules[j].Priority != nil {
+			priorityJ = *modules[j].Priority
+		}
+
+		return priorityI < priorityJ
+	})
+	return modules
+}
+
+// getMyModules returns all registered modules sorted by priority
+func (r *moduleRegistry) getMyModules() []MyModuleInfo {
+    r.mu.RLock()
+    defer r.mu.RUnlock()
+
+    modules := make([]MyModuleInfo, 0, len(r.modules))
+    for _, m := range r.modules {
+        modules = append(modules, m)
+    }
+    return sortMyModules(modules)
+}
+
+// coreModuleHooks manages module initialization and hook registration.
+type coreModuleHooks struct {
+	moduleInstances []MyModule
+	hookRegistry *hookRegistry
+	logger *zap.Logger
+}
+
+func newCoreModuleHooks(logger *zap.Logger) *coreModuleHooks {
+	return &coreModuleHooks{
+		hookRegistry: newHookRegistry(),
+		logger: logger,
+	}
+}
+
+// initCoreModuleHooks instantiates each module, provisions it,
+// registers any implemented hooks, and saves the hook registry.
+func (c *coreModuleHooks) initCoreModuleHooks(ctx context.Context, modules []MyModuleInfo) error {
+	hookRegistry := newHookRegistry()
+	var instances []MyModule
+
+	for _, info := range modules {
+		now := time.Now()
+		moduleInstance := info.New()
+
+		if err := moduleInstance.Provision(ctx); err != nil {
+			return fmt.Errorf("failed to provision module %s: %w", info.ID, err)
+		}
+
+		hookRegistry.AddApplicationLifecycle(moduleInstance)
+		hookRegistry.AddGraphQLServerLifecycle(moduleInstance)
+		hookRegistry.AddRouterLifecycle(moduleInstance)
+		hookRegistry.AddSubgraphLifecycle(moduleInstance)
+		hookRegistry.AddOperationLifecycle(moduleInstance)
+
+		c.logger.Info("Core Module System: Module registered",
+			zap.String("id", string(info.ID)),
+			zap.String("duration", time.Since(now).String()),
+		)
+
+		instances = append(instances, moduleInstance)
+	}
+
+	c.hookRegistry = hookRegistry
+	c.moduleInstances = instances
+
+	return nil
+}
+
+func (c *coreModuleHooks) cleanupCoreModuleHooks(ctx context.Context) error {
+	for _, moduleInstance := range c.moduleInstances {
+		if err := moduleInstance.Cleanup(); err != nil {
```

Comment:
`Cleanup` should also have a context so we can ensure that we don't wait forever e.g. the server shutdown enforces a max timeout when shutting down. We could forward this context here.